File: src/zeus-web/src/state/meter-display-settings-store.ts
// SPDX-License-Identifier: GPL-2.0-or-later
//
// Zeus — OpenHPSDR Protocol-1 / Protocol-2 client.
// Copyright (C) 2025-2026 Brian Keating (EI6LF) and contributors.
//
// See ATTRIBUTIONS.md at the repository root for the full provenance
// statement and per-component attribution.

import { create } from 'zustand';
import {
  fetchMeterDisplaySettings,
  updateMaxDisplayedWatts,
  updateSMeterOffsetDb,
  MAX_DISPLAYED_WATTS_DEFAULT,
  MAX_DISPLAYED_WATTS_MAX,
  MAX_DISPLAYED_WATTS_MIN,
  SMETER_OFFSET_DEFAULT_DB,
  SMETER_OFFSET_MIN_DB,
  SMETER_OFFSET_MAX_DB,
} from '../api/meter-display-settings';

// Display-side meter calibration knobs (GitHub #426). Persisted server-side
// in zeus-prefs.db (LiteDB) so the calibration follows the operator across
// browsers / devices. NO localStorage mirror per project_litedb_tx_filter
// _persistence / project_rotator_resume lessons.
//
// The component should call setSMeterOffsetDbLocal(...) on every keystroke
// (responsive display) and persistSMeterOffsetDb(...) on blur / Enter so
// the server only sees the committed value.

type MeterDisplaySettingsState = {
  sMeterOffsetDb: number;
  // 0 = "no override, use radio's MaxWatts as the meter full scale".
  maxDisplayedWatts: number;
  hydrated: boolean;
  setSMeterOffsetDbLocal: (dB: number) => void;
  persistSMeterOffsetDb: (dB: number) => Promise<void>;
  setMaxDisplayedWattsLocal: (w: number) => void;
  persistMaxDisplayedWatts: (w: number) => Promise<void>;
};

function clampOffset(v: number): number {
  if (!Number.isFinite(v)) return SMETER_OFFSET_DEFAULT_DB;
  if (v < SMETER_OFFSET_MIN_DB) return SMETER_OFFSET_MIN_DB;
  if (v > SMETER_OFFSET_MAX_DB) return SMETER_OFFSET_MAX_DB;
  return v;
}

function clampMaxWatts(v: number): number {
  if (!Number.isFinite(v)) return MAX_DISPLAYED_WATTS_DEFAULT;
  if (v <= 0) return MAX_DISPLAYED_WATTS_DEFAULT;
  if (v < MAX_DISPLAYED_WATTS_MIN) return MAX_DISPLAYED_WATTS_MIN;
  if (v > MAX_DISPLAYED_WATTS_MAX) return MAX_DISPLAYED_WATTS_MAX;
  return v;
}

export const useMeterDisplaySettingsStore = create<MeterDisplaySettingsState>((set) => ({
  sMeterOffsetDb: SMETER_OFFSET_DEFAULT_DB,
  maxDisplayedWatts: MAX_DISPLAYED_WATTS_DEFAULT,
  hydrated: false,
  setSMeterOffsetDbLocal: (dB) => {
    set({ sMeterOffsetDb: clampOffset(dB) });
  },
  persistSMeterOffsetDb: async (dB) => {
    const next = clampOffset(dB);
    set({ sMeterOffsetDb: next });
    try {
      const result = await updateSMeterOffsetDb(next);
      set({
        sMeterOffsetDb: result.sMeterOffsetDb,
        maxDisplayedWatts: result.maxDisplayedWatts,
      });
    } catch {
      // Network / server failure — keep the local value. Next commit
      // will retry. We deliberately do NOT roll back: the operator's
      // intent is the value they just dialed in.
    }
  },
  setMaxDisplayedWattsLocal: (w) => {
    set({ maxDisplayedWatts: clampMaxWatts(w) });
  },
  persistMaxDisplayedWatts: async (w) => {
    const next = clampMaxWatts(w);
    set({ maxDisplayedWatts: next });
    try {
      const result = await updateMaxDisplayedWatts(next);
      set({
        sMeterOffsetDb: result.sMeterOffsetDb,
        maxDisplayedWatts: result.maxDisplayedWatts,
      });
    } catch {
      // See above — keep the local value on failure.
    }
  },
}));

async function hydrateFromServer(): Promise<void> {
  try {
    const server = await fetchMeterDisplaySettings();
    useMeterDisplaySettingsStore.setState({
      sMeterOffsetDb: server.sMeterOffsetDb,
      maxDisplayedWatts: server.maxDisplayedWatts,
      hydrated: true,
    });
  } catch {
    // Backend unreachable — keep the defaults. Stay unhydrated so the
    // next successful fetch (or a PUT) populates the store.
    useMeterDisplaySettingsStore.setState({ hydrated: false });
  }
}

void hydrateFromServer();

File: src/Zeus.Server.Hosting/MeterDisplaySettingsStore.cs
// SPDX-License-Identifier: GPL-2.0-or-later
//
// Zeus — OpenHPSDR Protocol-1 / Protocol-2 client.
// Copyright (C) 2025-2026 Brian Keating (EI6LF) and contributors.
//
// See ATTRIBUTIONS.md at the repository root for the full provenance
// statement and per-component attribution.

using LiteDB;
using Zeus.Contracts;

namespace Zeus.Server;

// Per-operator display-side meter calibration knobs (GitHub #426).
// Same single-row LiteDB pattern as PanWfSplitStore / BottomPinStore.
//
//   * SMeterOffsetDb — signed dB offset added to the displayed RX dBm
//     value. Trim for real-world antenna / coax / preamp combinations
//     that shift the WDSP-internal reading by a few dB. Clamped ±20 dB.
//     Default 0.
//   * MaxDisplayedWatts — operator override for the TX forward-power
//     meter full-scale, in Watts. Lets an operator set e.g. 25 W on
//     a 100 W bracket so the indication fills the bar. Clamped
//     1..1000 W. 0 means "no override, use radio's MaxWatts as full
//     scale" (the historical behaviour). Default 0.
//
// Display-only — the underlying WDSP / radio physics are untouched.
public sealed class MeterDisplaySettingsStore : IDisposable
{
    public const double SMeterOffsetMinDb = -20.0;
    public const double SMeterOffsetMaxDb = 20.0;
    public const double SMeterOffsetDefaultDb = 0.0;

    public const double MaxDisplayedWattsMin = 1.0;
    public const double MaxDisplayedWattsMax = 1000.0;
    // 0 = "no override, use the radio's rated MaxWatts" — same value
    // the wire DTO surfaces to the frontend.
    public const double MaxDisplayedWattsDefault = 0.0;

    private readonly LiteDatabase _db;
    private readonly ILiteCollection<MeterDisplaySettingsEntry> _docs;
    private readonly ILogger<MeterDisplaySettingsStore> _log;
    private readonly object _sync = new();

    public event Action? Changed;

    public MeterDisplaySettingsStore(ILogger<MeterDisplaySettingsStore> log, string? dbPathOverride = null)
    {
        _log = log;
        var dbPath = dbPathOverride ?? PrefsDbPath.Get();
        var dir = Path.GetDirectoryName(dbPath);
        if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
        {
            Directory.CreateDirectory(dir);
        }

        _db = new LiteDatabase($"Filename={dbPath};Connection=shared");
        _docs = _db.GetCollection<MeterDisplaySettingsEntry>("meter_display_settings");

        _log.LogInformation("MeterDisplaySettingsStore initialized at {Path}", dbPath);
    }

    public MeterDisplaySettingsDto Get()
    {
        lock (_sync)
        {
            var e = _docs.FindAll().FirstOrDefault();
            if (e is null)
            {
                return new MeterDisplaySettingsDto(
                    SMeterOffsetDb: SMeterOffsetDefaultDb,
                    MaxDisplayedWatts: MaxDisplayedWattsDefault);
            }
            return new MeterDisplaySettingsDto(
                SMeterOffsetDb: ClampOffset(e.SMeterOffsetDb),
                MaxDisplayedWatts: ClampMaxWatts(e.MaxDisplayedWatts));
        }
    }

    public MeterDisplaySettingsDto SetSMeterOffsetDb(double offsetDb)
    {
        var clamped = ClampOffset(offsetDb);
        lock (_sync)
        {
            var e = _docs.FindAll().FirstOrDefault() ?? new MeterDisplaySettingsEntry();
            e.SMeterOffsetDb = clamped;
            e.UpdatedUtc = DateTime.UtcNow;
            if (e.Id == 0) _docs.Insert(e);
            else _docs.Update(e);
        }
        Changed?.Invoke();
        return Get();
    }

    public MeterDisplaySettingsDto SetMaxDisplayedWatts(double maxWatts)
    {
        var clamped = ClampMaxWatts(maxWatts);
        lock (_sync)
        {
            var e = _docs.FindAll().FirstOrDefault() ?? new MeterDisplaySettingsEntry();
            e.MaxDisplayedWatts = clamped;
            e.UpdatedUtc = DateTime.UtcNow;
            if (e.Id == 0) _docs.Insert(e);
            else _docs.Update(e);
        }
        Changed?.Invoke();
        return Get();
    }

    private static double ClampOffset(double v)
    {
        if (double.IsNaN(v) || double.IsInfinity(v)) return SMeterOffsetDefaultDb;
        if (v < SMeterOffsetMinDb) return SMeterOffsetMinDb;
        if (v > SMeterOffsetMaxDb) return SMeterOffsetMaxDb;
        return v;
    }

    private static double ClampMaxWatts(double v)
    {
        if (double.IsNaN(v) || double.IsInfinity(v)) return MaxDisplayedWattsDefault;
        // 0 (or any non-positive value) = "no override". Accept verbatim.
        // Anything else gets clamped into the legal [1, 1000] range.
        if (v <= 0) return MaxDisplayedWattsDefault;
        if (v < MaxDisplayedWattsMin) return MaxDisplayedWattsMin;
        if (v > MaxDisplayedWattsMax) return MaxDisplayedWattsMax;
        return v;
    }

    public void Dispose() => _db.Dispose();
}

public sealed class MeterDisplaySettingsEntry
{
    public int Id { get; set; }
    /// <summary>Signed dB offset applied to the displayed RX dBm.
    /// Clamped ±20 dB on read/write. LiteDB hydrates as 0.0 for older
    /// rows that pre-date this field, which matches the default.</summary>
    public double SMeterOffsetDb { get; set; }
    /// <summary>Operator override for the TX forward-power meter full
    /// scale, in Watts. 0 = "no override, use the radio's rated
    /// MaxWatts". Clamped [1, 1000] when non-zero. LiteDB hydrates as
    /// 0.0 for older rows that pre-date this field, which matches the
    /// default (no override).</summary>
    public double MaxDisplayedWatts { get; set; }
    public DateTime UpdatedUtc { get; set; }
}
